pyat/at/lattice/lattice_object.py

```diff
@@ -312,21 +301,125 @@ def _addition_filter(self, elems: Iterable[Element]):
                 self._cell_harmnumber = int(round(frequency / rev))
         self._radiation |= params.pop('_radiation')
 
-    def insert(self, idx: SupportsIndex, elem: Element):
+    def insert(self, idx: SupportsIndex, elem: Element, copy=False):
+        r"""This method allow to insert an AT element in the lattice.
+
+            Parameters:
+                idx (SupportsIndex): index at which the lement is inserted
+                elem (Element): AT element to be inserted in the lattice
+
+            Keyword Arguments:
+                copy(bool): Default :py:obj:`True`. If :py:obj:`True`
+                            a deep copy of elem is used
+        """
         # noinspection PyUnusedLocal
         # scan the new element to update it
-        elist = list(self._addition_filter([elem]))
+        elist = list(self._addition_filter([elem], copy=copy))
         super().insert(idx, elem)
 
-    def extend(self, elems: Iterable[Element]):
+    def extend(self, elems: Iterable[Element], copy=False):
+        r"""This method adds all the elements of `elems` to the end of the
+            lattice. The behavior is the same as for a :py:obj:`list`
+
+            Equivalents syntaxes:
+            >>>ring.extend(elems)
+            >>>ring += elems
+
+            Parameters:
+                elem (Iterable[Element]): Sequence of AT elements to be
+                                          appended to the lattice
+
+            Keyword Arguments:
+                copy(bool): Default :py:obj:`True`. If :py:obj:`True`
+                            deep copies of each element of elems are used
+        """
         if hasattr(self, '_energy'):
             # When unpickling a Lattice, extend is called before the lattice
             # is initialized. So skip this.
-            elems = self._addition_filter(elems)
+            elems = self._addition_filter(elems, copy=copy)
         super().extend(elems)
 
-    def append(self, elem: Element):
-        self.extend([elem])
+    def append(self, elem: Element, copy=False):
+        r"""This method overwrites the inherited method
+            :py:meth:`list.append()`,
+            it behavior is changed, it accepts only AT lattice elements
+            :py:obj:`Element` as input argument.
+
+            Equivalents syntaxes:
+            >>>ring.append(elem)
+            >>>ring += [elem]
```

Comment:
Missing space: `>>> ring.append(elem)`. (in many places)

Reply:
done